Stages as xblocks - part 2

e-kolpakov opened the pull request:
Contains the following:
* Sending completion events for basic and assessment stages
* Added Completion Stage - stage that needs to be explicitly marked as complete
* Support for activate_block_id navigation
* Static help XBlocks
* Lots of other stuff

bradenmacdonald commented on group_project_v2/templates/html/stages/completion.html:
```diff
@@ -0,0 +1,10 @@
+  <div class="group-project-completion-label">{{ completion_label}}</div>
+  <input type="checkbox" class="group-project-completion-checkmark" id="group-project-completion-checkmark"
+         {% if completed %}checked="'checked" disabled="disabled" {% endif %}>
+  <label for="group-project-completion-checkmark">{% trans "Yes, I have completed this task" %}</label>
```
Should this message be customizable?

e-kolpakov replied:
@bradenmacdonald they haven't asked for it (at least I didn't notice) and it would be pretty straightforward to add later.

bradenmacdonald commented on group_project_v2/utils.py:
```diff
@@ -95,4 +98,91 @@ def mean(value_array):
+    if not stages:
+        return None
+
+    if all(not stage.is_open for stage in stages):
```
If the block contains no stages, there are no validation warnings in Studio, but the LMS throws a 500 on this line: `AttributeError: 'NoneType' object has no attribute 'is_open'`. For some reason, `stages` is `[None]` in that case.

e-kolpakov replied:
@bradenmacdonald I thought I've fixed it - thank you for noticing that, I'll double check it.

bradenmacdonald commented:
@e-kolpakov I haven't really looked at the code (leaving that to @Kelketek); just did a quick smoke test as you requested. I'm not really sure how to test this - are there instructions somewhere? I did my best to check it out, and here's what I found:

1. Tests are not passing due to setup.py listing a different name for the XBlocks than the XML tags
2. I'm not sure if it is supposed to work in the LMS, but it does not:
   ![screen shot 2015-07-20 at 5 54 20 pm](https://cloud.githubusercontent.com/assets/945577/8790962/0c86883e-2f09-11e5-8332-26071ac57465.png)
   If it's not supposed to work in the LMS, can it display a message like "This only works in apros" ?
3. The font awesome icons aren't appearing in Apros. I'm on the `group_project_v2` branch:
    
![screen shot 2015-07-20 at 6 33 59 pm](https://cloud.githubusercontent.com/assets/945577/8791314/00af2494-2f0e-11e5-905f-5656e60b1afa.png)
4. When I click a resource link it just opens up a `i4x://` URL in my browser, which doesn't work.
5. When I mark a Completion Stage as complete, the navigator doesn't update and still shows it as incomplete:
   
![screen shot 2015-07-20 at 7 10 35 pm](https://cloud.githubusercontent.com/assets/945577/8791699/1e954006-2f13-11e5-9a2c-c7871086fbcc.png)
6. `jump_to_id` / `activate_block_id` seems to work well, but it's hard to get the block IDs in Studio. Can they be displayed somewhere?

e-kolpakov commented:
@bradenmacdonald 

* Tests - That's part of "Stabilization" task
* It's not supposed to work in LMS (at least yet) as Project Navigator is rendered in Apros only. @antoviaque not sure if it was agreed initially or have I just assumed that. Do we need to get it to work in LMS before deadline?
* icons: strange, never saw that. Have you rebuilt Apros static assets? 
* Resource links - local Apros/LMS installation is not configured to actually serve them. In production/QA they use amazon S3 I believe.
* I'm not sure if I implemented sending a signal to PN or not. To be honest, assessment and basic stages do not update PN status right away as well. I'll see what I can do here.
* Block ids - good point, will do.

antoviaque commented:
@e-kolpakov Yes, for now it's only used in the LMS - although a long-term goal is to get this to be usable by the community as a whole, who don't have apros. Since on the McKinsey instance the LMS is not accessible at all, I'm fine with leaving as is (not sure how easy it would be to display message in the LMS saying that it isn't supported there, since the LMS is used by apros to render the content), as long as we have a BIG HUGE warning on the git repository, so people don't try to install it unknowingly in the first place.

For the tests btw, it's a good occasion to mention that even in large projects like this I would really prefer if the tests were written progressively, ie each PR containing test coverage for the code it is adding, instead of being kept for the end. Otherwise we don't properly count the devlopment time on each of the individual stories, it's hard to ensure everything is properly tested when done in bulk at the end (both for you and the reviewer), and we'll definitely end up not covering some parts when the final deadlines are looming. Sven got caught on this too for Dalite.

e-kolpakov commented:
@antoviaque did you mean "Yes, for now it's only used in the Apros"?

As for tests - this time it was a conscious decision - if we did progressive test development the tests would significantly reduce the development speed, as the structure of the codebase changed quite significantly - even high-level integration tests needs some work after this conversion, and most unit tests would likely be thrown away. I agree with you about advantages of progressive test development, but this time it was my decision to delay tests until we have more or less stable final structure of the code. It might be a mistake, but we'll know it soon (i.e. if we meet the deadline it wasn't a mistake :)).

e-kolpakov commented:
@bradenmacdonald  I've double-checked completion stage state update - it is actually updated, even in your case - on the screenshot attached it says it has "incomplete" class. There are three stage states: 
* Not started - when no peers completed the stage
* Incomplete - some of peers completed the stage (maybe you)
* Complete - when all peers completed the stage( including you)

Having said that it looks like it might be a bit non-intuitive, as there're alternative view to that: mark as complete if current user is done with the stage, as incomplete when it's partially completed. I've asked a question on design document page about that - let's see what they answer.

antoviaque commented:
@ekolpakov Yes, I meant in apros.

And fair enough, it's true that there are scenarios where delaying tests makes sense, when the code base is meant to change significantly. The only thing in these cases in the future would be to make that decision explicitly, as part of the specs & tickets description writing, to ensure that it is a conscious decision for all parties (including the reviewer), and is accounted in the schedule? Btw this is not a reproach, more something which had been ambiguous on several large projects largely, so it was an occasion for me to bring it up for the future.

e-kolpakov commented:
@antoviaque sure, I didn't take it as reaproach - I still remember that in this case it was planned to be so from the beginning, at least it was included in Tech Spec as separate task.

bradenmacdonald commented:
@e-kolpakov I did rebuild the Apros static assets, so I have no idea why the font awesome icons aren't appearing.

All your other comments sound good - I'll leave the rest of the review to Jonathan.

Kelketek commented on group_project_v2/public/js/components/submission.js:
```diff
@@ -14,6 +16,24 @@ function GroupProjectSubmissionBlock(runtime, element) {
         $(document).trigger('group_project_v2.submission.upload_complete', uploadXHR);
     }
 
+    var message_box = $(".message"); // searching globally - not a typo: message box is created at group project level
```
Any chance there's a cleaner way to do this? I know it's unlikely to ever be an issue in practice, but it would be nice to be able to scope this somehow.

Kelketek commented on group_project_v2/public/js/components/video_resource.js:
```diff
@@ -0,0 +1,3 @@
+function VideoResourceXBlock(runtime, element) {
+
+}
```
Is there supposed to be something here?

Kelketek commented on group_project_v2/public/js/stages/completion.js:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    var $form = $(".group-project-completion-form", element);
+    var message_box = $(".message"); // searching globally - not a typo: message box is created at group project level
```
Same question here about scoping

Kelketek commented:
If I try to add a block to the Project navigator, I get stuck with a validation error "Project Navigator must contain Navigation view" and no way to add a Navigation block.

e-kolpakov commented:
@Kelketek I've fixed the issue with PN you had, addressed your other notes and also implemented latest updates to stage status criteria: basically they are now user-centered rather than group centered. TO be precise:

* Text, assessment and completion stages are marked as complete *for this user only* when user reviews it
* Review stages are marked as incomplete when user answers at least one required question and as complete when user answers all of them.
* Submission stage is marked as incomplete *for all users in group* when at least one submission is sent and as a complete when all submissions is sent.

Kelketek commented:
@e-kolpakov For the different stage types, can a specific display name be given for each by default? Right now it doesn't matter if I add a Peer Review or a submission stage-- their display names all read "Group Project V2 Stage"

e-kolpakov commented:
@Kelketek yes, give me a couple of minutes for that :)

Kelketek commented:
@e-kolpakov I managed to get this traceback when viewing the group work in Apros.
```
2015-07-22 10:19:25,234 ERROR 6684 [group_project_v2.project_navigator] project_navigator.py:71 - <class 'opaque_keys.edx.locator.BlockUsageLocator'>: 
Traceback (most recent call last):
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/project_navigator.py", line 66, in _get_activated_view_type
    usage_id = BlockUsageLocator.from_string(activate_block_id)
  File "/edx/app/edxapp/venvs/edxapp/src/opaque-keys/opaque_keys/__init__.py", line 191, in from_string
    return getattr(cls, 'deprecated_fallback')._from_deprecated_string(serialized)
  File "/edx/app/edxapp/venvs/edxapp/src/opaque-keys/opaque_keys/edx/locator.py", line 990, in _from_deprecated_string
    raise InvalidKeyError(BlockUsageLocator, serialized)
InvalidKeyError: <class 'opaque_keys.edx.locator.BlockUsageLocator'>: 
```

This doesn't happen when selecting a stage from the navigator, only when first clicking the group work button.

e-kolpakov commented:
@Kelketek that's just fine - if `activate_block_id` is empty it throws this error, intercepts and logs it, than continues to go on with default stage.

There is one other "normal" exception: if stage is marked as complete more than once (get 409 Conflict, logs and suppresses it)

Kelketek commented:
@e-kolpakov How about this one?

```
2015-07-22 10:41:01,501 ERROR 6558 [group_project_v2.mixins] mixins.py:60 - 'NoneType' object is not callable
Traceback (most recent call last):
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/mixins.py", line 58, in user_id
    return int(self.real_user_id(self.anonymous_student_id))
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/mixins.py", line 72, in real_user_id
    self._known_real_user_ids[anonymous_student_id] = self.runtime.get_real_user(anonymous_student_id).id
TypeError: 'NoneType' object is not callable
```

e-kolpakov commented:
@Kelketek that one should happen in studio and workbench as those runtimes does not provide `get_real_user` method. This particular method was mostly (if not entirely) preserved from V1, so it should remain the same, unless there is a better way to know real user id

Kelketek commented:
@e-kolpakov Got this one when trying to work through the stages. Caused Apros to throw a 500:

```
2015-07-22 10:50:22,320 ERROR 6684 [django.request] base.py:213 - Internal Server Error: /api/server/courses/group/test/woo/navigation/4cc27e1ca6fa452f9ecc2eb7e4cf59ac
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 109, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/compat.py", line 137, in view
    return self.dispatch(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 77, in wrapped_view
    return view_func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 400, in dispatch
    response = self.handle_exception(exc)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 397, in dispatch
    response = handler(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/api_manager/courses/views.py", line 1994, in get
    (course_key, chapter, section, vertical, position, final_target_id) = path_to_location(modulestore(), usage_key)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/search.py", line 77, in path_to_location
    raise NoPathToItem(usage_key)
NoPathToItem: i4x://group/test/group-project-v2-stage-peer-review/4cc27e1ca6fa452f9ecc2eb7e4cf59ac
```

e-kolpakov commented:
@Kelketek you must have more than one block with "4cc27e1ca6fa452f9ecc2eb7e4cf59ac" id. Basically, export-import preserves `block_id` part of block usage locator, so it is possible to have blocks with same `block_id` in different courses (as mongos key is block_id + course_id + some_other_stuff). When used jump_to_id to such block id is performed `path_to_location` might take pick block from other course and than fail to find it in current course. 

It's a bug in `path_to_location`

Kelketek commented:
@e-kolpakov In the project navigator, if I click on the simple resource, the URL is a block ID, and my browser tries to load 'i4x://test/v2/gp-v2-resource/7b82889c97064d45a858612ba750dc44'

Kelketek commented:
@e-kolpakov Should it be 'Evaluation: Group Assessment' instead of 'Grade: Group Assessment'?

Also, @bradenmacdonald seems to have had the same issue I've had with the fonts. Since both of us are having it and you aren't, I'd really like you to look into this. Could you clear out all static assets you've generated for Apros, and then recollect them from scratch? If there are any left over files, that might allow it.

If there are any directives for the NginX configuration that are missing from Apros's setup docs which might be responsible, those would be good to update. :)

e-kolpakov commented:
@Kelketek I've tried clearing up assets, and switching to other nginx rule sets - basically it didn't help to reproduce the problem. If that matters, I use `mcka_apros_production` set, font-awesome is loaded from http://apros.mcka.local/static/css/font-awesome.css. `fa-files-o` (resources view selector icon) is located at line 721.

So, I've went to QA and set up a [demo course](https://qa.mckinsey.edx.org/courses/Opencraft/GP2_1/T1/group_work). Login: AliceDoe, standard password. Icons are there, so it must be a problem on your side. Demo runs on part 1 version (since part 2 is not yet merged) and latest master (which is on par with part 1), so there's a tiny chance that I've blown something, but I can't think of anything that could cause such an effect.

Kelketek commented:
@e-kolpakov New traceback:

```
Traceback (most recent call last):
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/group_project.py", line 77, in _get_activity_to_display
    usage_id = BlockUsageLocator.from_string(target_stage_id)
  File "/edx/app/edxapp/venvs/edxapp/src/opaque-keys/opaque_keys/__init__.py", line 191, in from_string
    return getattr(cls, 'deprecated_fallback')._from_deprecated_string(serialized)
  File "/edx/app/edxapp/venvs/edxapp/src/opaque-keys/opaque_keys/edx/locator.py", line 990, in _from_deprecated_string
    raise InvalidKeyError(BlockUsageLocator, serialized)
InvalidKeyError: <class 'opaque_keys.edx.locator.BlockUsageLocator'>: 
```

Kelketek commented:
@e-kolpakov @bradenmacdonald  I've found what seems to fix the font problem for me. You need to add the `fa` class to any span that's going to use a font-awesome icon. For instance:

```
<span class="group-project-stage-state completed"></span>
```
...should be...

```
<span class="fa group-project-stage-state completed"></span>
```

e-kolpakov commented:
@Kelketek we're in a hurry to deploy it at least to QA, so I'm merging this without your explicit +1. I've implemented `fa` fix you requested... and a bunch of other things. You still need to review those and give your +1, ut all the further fixes will happen as a separate PR (I hope there wouldn't be any)

Kelketek commented:
```
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 109, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 1017, in xblock_view
    fragment = instance.render(view_name, context=request.GET)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/core.py", line 155, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1770, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1140, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/project_navigator.py", line 85, in student_view
    child_fragment = view.render('student_view', context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/core.py", line 155, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1770, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1140, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/project_navigator.py", line 290, in student_view
    activity_fragment = activity.render("navigation_view", context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/core.py", line 155, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1770, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1140, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/src/xblock/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/utils.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/group_project.py", line 282, in navigation_view
    for stage in self.available_stages:
  File "/edx/app/edxapp/xblocks/xblock-group-project-v2/group_project_v2/group_project.py", line 215, in available_stages
    if stage.available_to_current_user:
AttributeError: 'HiddenDescriptorWithMixins' object has no attribute 'available_to_current_user'
[24/Jul/2015 12:38:24] "GET /courses/test/v2/group/xblock/i4x:;_;_test;_v2;_gp-v2-navigator;_220a3ca1e3624ad1a70c0d1600778cc9/view/student_view?activate_block_id=&_=1437755899161 HTTP/1.0" 500 62888
```

This is keeping groupwork from loading at all.

Kelketek commented:
@e-kolpakov The fix you mentioned worked!

Some of the icons are showing now. You did not quite get them all, though. The blue buttons on the bottom don't have the class, so only "TA" shows up right, and the completion markers are also missing the right class.
